File: examples/25-caching/python/cache-service/main.py
import asyncio
import json
import logging
import os
import time
from contextlib import asynccontextmanager
from typing import Optional

import asyncpg
import redis.asyncio as aioredis
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel

log = logging.getLogger("cache-service")
logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")

DATABASE_URL = os.getenv("DATABASE_URL", "postgres://appuser:apppass@postgres:5432/appdb")
REDIS_URL = os.getenv("REDIS_URL", "redis://redis:6379")
TTL = 60

pool: asyncpg.Pool
cache: aioredis.Redis


class ProductIn(BaseModel):
    name: str
    price_cents: int


class EventIn(BaseModel):
    id: str
    type: str
    payload: dict = {}


class MetricIn(BaseModel):
    value: float
    tags: dict = {}


# ---------------------------------------------------------------------------
# Read-through wrapper
# ---------------------------------------------------------------------------
class ReadThrough:
    def __init__(self, redis_client, loader, ttl=60, prefix=""):
        self.redis = redis_client
        self.loader = loader
        self.ttl = ttl
        self.prefix = prefix

    async def get(self, id: str) -> Optional[dict]:
        key = f"{self.prefix}{id}"
        cached = await self.redis.get(key)
        if cached:
            return json.loads(cached)
        value = await self.loader(id)
        if value is not None:
            await self.redis.setex(key, self.ttl, json.dumps(value))
        return value


# ---------------------------------------------------------------------------
# Background tasks
# ---------------------------------------------------------------------------
FLUSH_BATCH = 100
FLUSH_PERIOD = 1.0
REFRESH_BEFORE_S = 10


async def flusher():
    """Write-back flusher: drain dirty set to Postgres in batches."""
    while True:
        await asyncio.sleep(FLUSH_PERIOD)
        try:
            ids = await cache.spop("metric:dirty", FLUSH_BATCH)
            if not ids:
                continue
            async with pool.acquire() as conn:
                for mid_b in ids:
                    mid = mid_b if isinstance(mid_b, str) else mid_b.decode()
                    data = await cache.hgetall(f"metric:{mid}")
                    if data:
                        payload = {
                            (k if isinstance(k, str) else k.decode()): (
                                v if isinstance(v, str) else v.decode()
                            )
                            for k, v in data.items()
                        }
                        await conn.execute(
                            "INSERT INTO metrics (id, payload) VALUES ($1, $2) "
                            "ON CONFLICT (id) DO UPDATE SET payload = $2, ts = NOW()",
                            mid,
                            json.dumps(payload),
                        )
                log.info("flusher: persisted %d metrics", len(ids))
        except Exception:
            log.exception("flusher error")


async def refresher():
    """Refresh-ahead: re-warm hot keys before their TTL fires."""
    while True:
        await asyncio.sleep(5)
        try:
            cutoff = time.time() - 300
            await cache.zremrangebyscore("product:hot", 0, cutoff)
            hot_ids = await cache.zrange("product:hot", 0, -1)
            for pid_b in hot_ids:
                pid = pid_b if isinstance(pid_b, str) else pid_b.decode()
                ttl_val = await cache.ttl(f"ra:product:{pid}")
                if 0 < ttl_val < REFRESH_BEFORE_S:
                    async with pool.acquire() as conn:
                        row = await conn.fetchrow(
                            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
                        )
                        if row:
                            await cache.setex(
                                f"ra:product:{pid}",
                                TTL,
                                json.dumps(dict(row)),
                            )
                            log.info("refresher: pre-warmed %s", pid)
        except Exception:
            log.exception("refresher error")


# ---------------------------------------------------------------------------
# App lifecycle
# ---------------------------------------------------------------------------
@asynccontextmanager
async def lifespan(app: FastAPI):
    global pool, cache
    pool = await asyncpg.create_pool(DATABASE_URL, min_size=2, max_size=10)
    cache = aioredis.from_url(REDIS_URL, decode_responses=False)

    flush_task = asyncio.create_task(flusher())
    refresh_task = asyncio.create_task(refresher())

    yield

    flush_task.cancel()
    refresh_task.cancel()
    await cache.aclose()
    await pool.close()


app = FastAPI(lifespan=lifespan)


# ---------------------------------------------------------------------------
# Helper: try cache, fall back to DB on Redis failure
# ---------------------------------------------------------------------------
async def safe_cache_get(key: str) -> Optional[bytes]:
    try:
        return await cache.get(key)
    except Exception:
        return None


async def safe_cache_set(key: str, value: str, ttl: int = TTL):
    try:
        await cache.setex(key, ttl, value)
    except Exception:
        pass


async def safe_cache_delete(key: str):
    try:
        await cache.delete(key)
    except Exception:
        pass


# ---------------------------------------------------------------------------
# 1. Cache-aside
# ---------------------------------------------------------------------------
@app.get("/cache-aside/products/{pid}")
async def cache_aside_get(pid: str):
    key = f"ca:product:{pid}"
    cached = await safe_cache_get(key)
    if cached:
        return {"source": "cache", **json.loads(cached)}

    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
        )
    if not row:
        raise HTTPException(404, "not found")
    data = dict(row)
    await safe_cache_set(key, json.dumps(data))
    return {"source": "db", **data}


@app.put("/cache-aside/products/{pid}")
async def cache_aside_update(pid: str, body: ProductIn):
    async with pool.acquire() as conn:
        await conn.execute(
            "UPDATE products SET name=$1, price_cents=$2 WHERE id=$3",
            body.name,
            body.price_cents,
            pid,
        )
    await safe_cache_delete(f"ca:product:{pid}")
    return {"ok": True, "pattern": "cache-aside", "action": "invalidated"}


# ---------------------------------------------------------------------------
# 2. Read-through
# ---------------------------------------------------------------------------
async def _load_product(pid: str) -> Optional[dict]:
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
        )
    return dict(row) if row else None


@app.get("/read-through/products/{pid}")
async def read_through_get(pid: str):
    rt = ReadThrough(cache, _load_product, ttl=TTL, prefix="rt:product:")
    result = await rt.get(pid)
    if result is None:
        raise HTTPException(404, "not found")

    key = f"rt:product:{pid}"
    cached = await safe_cache_get(key)
    source = "cache" if cached else "db"
    return {"source": source, **result}


# ---------------------------------------------------------------------------
# 3. Write-through
# ---------------------------------------------------------------------------
@app.get("/write-through/products/{pid}")
async def write_through_get(pid: str):
    key = f"wt:product:{pid}"
    cached = await safe_cache_get(key)
    if cached:
        return {"source": "cache", **json.loads(cached)}

    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
        )
    if not row:
        raise HTTPException(404, "not found")
    data = dict(row)
    await safe_cache_set(key, json.dumps(data))
    return {"source": "db", **data}


@app.put("/write-through/products/{pid}")
async def write_through_update(pid: str, body: ProductIn):
    async with pool.acquire() as conn:
        await conn.execute(
            "UPDATE products SET name=$1, price_cents=$2 WHERE id=$3",
            body.name,
            body.price_cents,
            pid,
        )
        row = await conn.fetchrow(
            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
        )
    await safe_cache_set(f"wt:product:{pid}", json.dumps(dict(row)))
    return {"ok": True, "pattern": "write-through", "action": "set"}


# ---------------------------------------------------------------------------
# 4. Write-around
# ---------------------------------------------------------------------------
@app.post("/write-around/events")
async def write_around_create(body: EventIn):
    async with pool.acquire() as conn:
        await conn.execute(
            "INSERT INTO events (id, type, payload) VALUES ($1, $2, $3)",
            body.id,
            body.type,
            json.dumps(body.payload),
        )
    return {"ok": True, "pattern": "write-around", "action": "db-only"}


@app.get("/write-around/events/{eid}")
async def write_around_get(eid: str):
    key = f"wa:event:{eid}"
    cached = await safe_cache_get(key)
    if cached:
        return {"source": "cache", **json.loads(cached)}

    async with pool.acquire() as conn:
        row = await conn.fetchrow("SELECT id, type, payload FROM events WHERE id=$1", eid)
    if not row:
        raise HTTPException(404, "not found")
    data = {"id": row["id"], "type": row["type"], "payload": json.loads(row["payload"])}
    await safe_cache_set(key, json.dumps(data))
    return {"source": "db", **data}


# ---------------------------------------------------------------------------
# 5. Write-back (write-behind)
# ---------------------------------------------------------------------------
@app.put("/write-back/metrics/{mid}")
async def write_back_write(mid: str, body: MetricIn):
    mapping = {"value": str(body.value), "tags": json.dumps(body.tags)}
    try:
        await cache.hset(f"metric:{mid}", mapping=mapping)
        await cache.sadd("metric:dirty", mid)
    except Exception:
        raise HTTPException(503, "cache unavailable")
    return {"ok": True, "pattern": "write-back", "action": "cached-for-flush"}


@app.get("/write-back/metrics/{mid}")
async def write_back_get(mid: str):
    try:
        data = await cache.hgetall(f"metric:{mid}")
        if data:
            payload = {
                (k if isinstance(k, str) else k.decode()): (
                    v if isinstance(v, str) else v.decode()
                )
                for k, v in data.items()
            }
            return {"source": "cache", "id": mid, **payload}
    except Exception:
        pass

    async with pool.acquire() as conn:
        row = await conn.fetchrow("SELECT id, payload FROM metrics WHERE id=$1", mid)
    if not row:
        raise HTTPException(404, "not found")
    return {"source": "db", "id": row["id"], "payload": json.loads(row["payload"])}


@app.get("/write-back/flush-status")
async def write_back_flush_status():
    try:
        dirty_count = await cache.scard("metric:dirty")
    except Exception:
        dirty_count = -1
    async with pool.acquire() as conn:
        db_count = await conn.fetchval("SELECT count(*) FROM metrics")
    return {"dirty_keys": dirty_count, "persisted_rows": db_count}


# ---------------------------------------------------------------------------
# 6. Refresh-ahead
# ---------------------------------------------------------------------------
@app.get("/refresh-ahead/products/{pid}")
async def refresh_ahead_get(pid: str):
    key = f"ra:product:{pid}"
    try:
        await cache.zadd("product:hot", {pid: time.time()})
    except Exception:
        pass

    cached = await safe_cache_get(key)
    if cached:
        return {"source": "cache", **json.loads(cached)}

    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT id, name, price_cents FROM products WHERE id=$1", pid
        )
    if not row:
        raise HTTPException(404, "not found")
    data = dict(row)
    await safe_cache_set(key, json.dumps(data))
    return {"source": "db", **data}


# ---------------------------------------------------------------------------
# Health
# ---------------------------------------------------------------------------
@app.get("/healthz")
async def healthz():
    return {"status": "ok"}


@app.get("/cache-keys")
async def cache_keys():
    """Debug endpoint: list all cache keys."""
    try:
        keys = []
        async for key in cache.scan_iter("*"):
            k = key if isinstance(key, str) else key.decode()
            keys.append(k)
        return {"keys": sorted(keys)}
    except Exception:
        return {"keys": [], "error": "cache unavailable"}

File: examples/25-caching/cpp/cache-service/src/main.cpp
#include <drogon/drogon.h>
#include <sw/redis++/redis++.h>
#include <libpq-fe.h>
#include <atomic>
#include <thread>
#include <chrono>
#include <mutex>
#include <sstream>
#include <optional>

using namespace drogon;
using namespace sw::redis;

// Global state
std::unique_ptr<Redis> g_redis;
PGconn* g_pg = nullptr;
std::atomic<int> g_persisted_rows{0};
std::atomic<bool> g_shutdown{false};
std::mutex g_redis_mutex;

// Helper: execute SQL query and return JSON result
Json::Value query_db(const std::string& query) {
    PGresult* res = PQexec(g_pg, query.c_str());
    Json::Value result;

    if (PQresultStatus(res) == PGRES_TUPLES_OK && PQntuples(res) > 0) {
        result["id"] = PQgetvalue(res, 0, 0);
        result["name"] = PQgetvalue(res, 0, 1);
        result["price_cents"] = std::stoi(PQgetvalue(res, 0, 2));
    }

    PQclear(res);
    return result;
}

// Helper: query event from DB
Json::Value query_event_db(const std::string& id) {
    std::string query = "SELECT id, type, payload FROM events WHERE id = '" + id + "'";
    PGresult* res = PQexec(g_pg, query.c_str());
    Json::Value result;

    if (PQresultStatus(res) == PGRES_TUPLES_OK && PQntuples(res) > 0) {
        result["id"] = PQgetvalue(res, 0, 0);
        result["type"] = PQgetvalue(res, 0, 1);

        Json::Reader reader;
        std::string payload_str = PQgetvalue(res, 0, 2);
        Json::Value payload;
        if (reader.parse(payload_str, payload)) {
            result["payload"] = payload;
        }
    }

    PQclear(res);
    return result;
}

// Helper: query metric from DB
Json::Value query_metric_db(const std::string& id) {
    std::string query = "SELECT id, payload FROM metrics WHERE id = '" + id + "'";
    PGresult* res = PQexec(g_pg, query.c_str());
    Json::Value result;

    if (PQresultStatus(res) == PGRES_TUPLES_OK && PQntuples(res) > 0) {
        result["id"] = PQgetvalue(res, 0, 0);

        Json::Reader reader;
        std::string payload_str = PQgetvalue(res, 0, 1);
        Json::Value payload;
        if (reader.parse(payload_str, payload)) {
            result["payload"] = payload;
        }
    }

    PQclear(res);
    return result;
}

// Helper: safe Redis GET
std::optional<std::string> redis_get(const std::string& key) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return std::nullopt;
        auto val = g_redis->get(key);
        return val;
    } catch (...) {
        return std::nullopt;
    }
}

// Helper: safe Redis SET
bool redis_set(const std::string& key, const std::string& value, std::chrono::seconds ttl = std::chrono::seconds(60)) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return false;
        g_redis->set(key, value, ttl);
        return true;
    } catch (...) {
        return false;
    }
}

// Helper: safe Redis DEL
bool redis_del(const std::string& key) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return false;
        g_redis->del(key);
        return true;
    } catch (...) {
        return false;
    }
}

// Helper: safe Redis SADD
bool redis_sadd(const std::string& set, const std::string& member) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return false;
        g_redis->sadd(set, member);
        return true;
    } catch (...) {
        return false;
    }
}

// Helper: safe Redis SREM
bool redis_srem(const std::string& set, const std::string& member) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return false;
        g_redis->srem(set, member);
        return true;
    } catch (...) {
        return false;
    }
}

// Helper: safe Redis SMEMBERS
std::vector<std::string> redis_smembers(const std::string& set) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return {};
        std::vector<std::string> members;
        g_redis->smembers(set, std::back_inserter(members));
        return members;
    } catch (...) {
        return {};
    }
}

// Helper: safe Redis KEYS
std::vector<std::string> redis_keys(const std::string& pattern) {
    try {
        std::lock_guard<std::mutex> lock(g_redis_mutex);
        if (!g_redis) return {};
        std::vector<std::string> keys;
        g_redis->keys(pattern, std::back_inserter(keys));
        return keys;
    } catch (...) {
        return {};
    }
}

// Background flusher thread
void writeback_flusher() {
    while (!g_shutdown) {
        std::this_thread::sleep_for(std::chrono::seconds(2));

        auto dirty_keys = redis_smembers("writeback:dirty");

        for (const auto& key : dirty_keys) {
            auto val = redis_get(key);
            if (!val) continue;

            // Parse the cached metric
            Json::Reader reader;
            Json::Value cached;
            if (!reader.parse(*val, cached)) continue;

            // Extract metric ID from key (metric:m1 -> m1)
            std::string id = key.substr(key.find(':') + 1);

            // Persist to DB
            Json::FastWriter writer;
            std::string payload = writer.write(cached["payload"]);

            std::ostringstream query;
            query << "INSERT INTO metrics (id, payload) VALUES ('"
                  << id << "', '" << payload << "') "
                  << "ON CONFLICT (id) DO UPDATE SET payload = EXCLUDED.payload";

            PGresult* res = PQexec(g_pg, query.str().c_str());
            if (PQresultStatus(res) == PGRES_COMMAND_OK) {
                g_persisted_rows++;
                redis_srem("writeback:dirty", key);
            }
            PQclear(res);
        }
    }
}

int main() {
    // Connect to Postgres
    const char* pg_conninfo = std::getenv("PG_CONNINFO");
    if (!pg_conninfo) pg_conninfo = "postgresql://appuser:apppass@localhost:5432/appdb";

    g_pg = PQconnectdb(pg_conninfo);
    if (PQstatus(g_pg) != CONNECTION_OK) {
        LOG_ERROR << "Postgres connection failed: " << PQerrorMessage(g_pg);
        return 1;
    }
    LOG_INFO << "Connected to Postgres";

    // Connect to Redis
    const char* redis_url = std::getenv("REDIS_URL");
    if (!redis_url) redis_url = "redis://localhost:6379";

    try {
        g_redis = std::make_unique<Redis>(redis_url);
        LOG_INFO << "Connected to Redis";
    } catch (const std::exception& e) {
        LOG_WARN << "Redis connection failed: " << e.what() << " (continuing without cache)";
    }

    // Start background flusher
    std::thread flusher(writeback_flusher);

    app().setUploadPath("/tmp");

    // --- healthz ---
    app().registerHandler("/healthz",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value resp;
            resp["status"] = "ok";
            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // 1. Cache-aside
    // ======================================================================

    // GET /cache-aside/products/:id
    app().registerHandler("/cache-aside/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("product:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Cache miss - query DB
            std::string query = "SELECT id, name, price_cents FROM products WHERE id = '" + id + "'";
            resp = query_db(query);

            if (!resp.empty()) {
                resp["source"] = "db";

                // Populate cache
                Json::FastWriter writer;
                Json::Value cache_val = resp;
                cache_val.removeMember("source");
                redis_set("product:" + id, writer.write(cache_val));
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // PUT /cache-aside/products/:id
    app().registerHandler("/cache-aside/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            auto json = req->getJsonObject();
            if (!json) {
                auto resp = HttpResponse::newHttpResponse();
                resp->setStatusCode(k400BadRequest);
                callback(resp);
                return;
            }

            std::string name = (*json)["name"].asString();
            int price = (*json)["price_cents"].asInt();

            std::ostringstream query;
            query << "UPDATE products SET name = '" << name
                  << "', price_cents = " << price
                  << " WHERE id = '" << id << "'";

            PGresult* res = PQexec(g_pg, query.str().c_str());
            PQclear(res);

            // Invalidate cache
            redis_del("product:" + id);

            auto resp = HttpResponse::newHttpResponse();
            resp->setStatusCode(k200OK);
            callback(resp);
        },
        {Put});

    // ======================================================================
    // 2. Read-through
    // ======================================================================

    // GET /read-through/products/:id
    app().registerHandler("/read-through/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("product:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Transparent cache layer loads from DB
            std::string query = "SELECT id, name, price_cents FROM products WHERE id = '" + id + "'";
            resp = query_db(query);

            if (!resp.empty()) {
                resp["source"] = "db";

                // Auto-populate cache
                Json::FastWriter writer;
                Json::Value cache_val = resp;
                cache_val.removeMember("source");
                redis_set("product:" + id, writer.write(cache_val));
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // 3. Write-through
    // ======================================================================

    // PUT /write-through/products/:id
    app().registerHandler("/write-through/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            auto json = req->getJsonObject();
            if (!json) {
                auto resp = HttpResponse::newHttpResponse();
                resp->setStatusCode(k400BadRequest);
                callback(resp);
                return;
            }

            std::string name = (*json)["name"].asString();
            int price = (*json)["price_cents"].asInt();

            // Write to DB
            std::ostringstream query;
            query << "UPDATE products SET name = '" << name
                  << "', price_cents = " << price
                  << " WHERE id = '" << id << "'";

            PGresult* res = PQexec(g_pg, query.str().c_str());
            PQclear(res);

            // Write to cache synchronously
            Json::Value cache_val;
            cache_val["id"] = id;
            cache_val["name"] = name;
            cache_val["price_cents"] = price;

            Json::FastWriter writer;
            redis_set("product:" + id, writer.write(cache_val));

            auto resp = HttpResponse::newHttpResponse();
            resp->setStatusCode(k200OK);
            callback(resp);
        },
        {Put});

    // GET /write-through/products/:id
    app().registerHandler("/write-through/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("product:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Fallback to DB
            std::string query = "SELECT id, name, price_cents FROM products WHERE id = '" + id + "'";
            resp = query_db(query);
            resp["source"] = "db";

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // 4. Write-around
    // ======================================================================

    // POST /write-around/events
    app().registerHandler("/write-around/events",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback) {
            auto json = req->getJsonObject();
            if (!json) {
                auto resp = HttpResponse::newHttpResponse();
                resp->setStatusCode(k400BadRequest);
                callback(resp);
                return;
            }

            std::string id = (*json)["id"].asString();
            std::string type = (*json)["type"].asString();
            Json::FastWriter writer;
            std::string payload = writer.write((*json)["payload"]);

            // Write to DB only (skip cache)
            std::ostringstream query;
            query << "INSERT INTO events (id, type, payload) VALUES ('"
                  << id << "', '" << type << "', '" << payload << "')";

            PGresult* res = PQexec(g_pg, query.str().c_str());
            PQclear(res);

            auto resp = HttpResponse::newHttpResponse();
            resp->setStatusCode(k200OK);
            callback(resp);
        },
        {Post});

    // GET /write-around/events/:id
    app().registerHandler("/write-around/events/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("event:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Query DB
            resp = query_event_db(id);

            if (!resp.empty()) {
                resp["source"] = "db";

                // Populate cache on read
                Json::FastWriter writer;
                Json::Value cache_val = resp;
                cache_val.removeMember("source");
                redis_set("event:" + id, writer.write(cache_val));
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // 5. Write-back (write-behind)
    // ======================================================================

    // PUT /write-back/metrics/:id
    app().registerHandler("/write-back/metrics/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            auto json = req->getJsonObject();
            if (!json) {
                auto resp = HttpResponse::newHttpResponse();
                resp->setStatusCode(k400BadRequest);
                callback(resp);
                return;
            }

            // Write to cache immediately
            Json::Value cache_val;
            cache_val["id"] = id;
            cache_val["payload"] = *json;

            Json::FastWriter writer;
            redis_set("metric:" + id, writer.write(cache_val));

            // Mark as dirty for background flush
            redis_sadd("writeback:dirty", "metric:" + id);

            auto resp = HttpResponse::newHttpResponse();
            resp->setStatusCode(k200OK);
            callback(resp);
        },
        {Put});

    // GET /write-back/metrics/:id
    app().registerHandler("/write-back/metrics/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("metric:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Fallback to DB
            resp = query_metric_db(id);
            if (!resp.empty()) {
                resp["source"] = "db";
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // GET /write-back/flush-status
    app().registerHandler("/write-back/flush-status",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value resp;
            resp["persisted_rows"] = g_persisted_rows.load();
            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // 6. Refresh-ahead
    // ======================================================================

    // GET /refresh-ahead/products/:id
    app().registerHandler("/refresh-ahead/products/{id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& id) {
            Json::Value resp;

            // Try cache first
            auto cached = redis_get("product:hot:" + id);
            if (cached) {
                Json::Reader reader;
                if (reader.parse(*cached, resp)) {
                    resp["source"] = "cache";
                    auto http_resp = HttpResponse::newHttpJsonResponse(resp);
                    callback(http_resp);
                    return;
                }
            }

            // Query DB
            std::string query = "SELECT id, name, price_cents FROM products WHERE id = '" + id + "'";
            resp = query_db(query);

            if (!resp.empty()) {
                resp["source"] = "db";

                // Populate cache with proactive refresh
                Json::FastWriter writer;
                Json::Value cache_val = resp;
                cache_val.removeMember("source");
                redis_set("product:hot:" + id, writer.write(cache_val), std::chrono::seconds(300));
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // ======================================================================
    // Cache keys endpoint
    // ======================================================================

    // GET /cache-keys
    app().registerHandler("/cache-keys",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value resp(Json::arrayValue);

            auto product_keys = redis_keys("product:*");
            for (const auto& key : product_keys) {
                resp.append(key);
            }

            auto event_keys = redis_keys("event:*");
            for (const auto& key : event_keys) {
                resp.append(key);
            }

            auto http_resp = HttpResponse::newHttpJsonResponse(resp);
            callback(http_resp);
        },
        {Get});

    // Start server
    LOG_INFO << "Starting cache-service on :8080";
    app().addListener("0.0.0.0", 8080).run();

    // Cleanup
    g_shutdown = true;
    flusher.join();
    PQfinish(g_pg);

    return 0;
}
